src/Test/QuickCheck/StateMachineModel.hs

```diff
@@ -36,26 +40,58 @@ import           Text.PrettyPrint.ANSI.Leijen            (Pretty, align, dot,
 
 import           Test.QuickCheck.StateMachineModel.Utils
 
+import           Unsafe.Coerce
+
 ------------------------------------------------------------------------
 
+liftSemSeq
+  :: forall cmd resp ix ref m
+  .  (Monad m, Enum ix, Typeable ref, RefKit cmd)
+  => Typeable resp
+  => Ord ix
+  => Show resp
+  => (cmd resp ref -> m resp)
+  -> cmd resp ix -> StateT (Map ix ref) m resp
+liftSemSeq sem cmd = do
+
+  env <- get
+  Untyped cmd' <- return $ fmap (\ix -> env M.! ix) $ Untyped cmd
+  let Just (cmd'' :: cmd resp ref) = ccast cmd'
+
+  case returnsRef' cmd'' of
+    Just Refl -> do
+      ref <- lift $ sem cmd''
+      let ix :: ix
+          ix = toEnum . length . M.keys $ env
+      modify $ M.insert ix ref
+      return $ unsafeCoerce ix
+    Nothing -> do
+      resp <- lift $ sem cmd''
+      return resp
+
 liftSem
   :: forall cmd resp ix ref m
   .  (Monad m, Enum ix, Typeable ref, RefKit cmd)
   => Typeable resp
+  => Ord (ix, Int)
   => Show resp
   => (cmd resp ref -> m resp)
-  -> cmd resp ix -> StateT [ref] m resp
-liftSem sem cmd = do
+  -> Int
+  -> cmd resp (ix, Int) -> StateT (Map (ix, Int) ref) m resp
+liftSem sem pid cmd = do
 
   env <- get
-  Untyped cmd' <- return $ fmap (\ix -> env !! fromEnum ix) $ Untyped cmd
+  Untyped cmd' <- return $ fmap (\ix -> env M.! ix) $ Untyped cmd
   let Just (cmd'' :: cmd resp ref) = ccast cmd'
 
   case returnsRef' cmd'' of
     Just Refl -> do
       ref <- lift $ sem cmd''
-      modify (++ [ ref ])
-      return ref
+      let ix :: ix
+          ix = toEnum . length . filter ((== pid) . snd) . M.keys $ env
```

Comment:
Maybe having the variables not start from zero again would be a better idea? Might need to do less shifts then.